kornia/augmentation/random_generator/_2d/salt_pepper_noise.py

```diff
@@ -0,0 +1,93 @@
+                with a shape of (B,).
+            - mask_salt: Binary masks (float) indicating the presence of Salt noise with a shape of (B, C, H, W).
+            - mask_pepper: Binary masks (float) indicating the presence of Pepper noise with a shape of (B, C, H, W).
+
+    Note:
+        The generated random numbers are not reproducible across different devices and dtypes. By default,
+        the parameters will be generated on CPU. This can be changed by calling
+        ``self.set_rng_device_and_dtype(device="cuda", dtype=torch.float64)``.
+    """
+
+    def __init__(
+        self,
+        amount: tuple[float, float],
+        salt_and_pepper: tuple[float, float],
+    ) -> None:
+        super().__init__()
+        self.amount = amount
+        self.salt_and_pepper = salt_and_pepper
+
+    def __repr__(self) -> str:
+        r"""Return a string representation of the object."""
+        repr = f"amount={self.amount}, salt_and_pepper={self.salt_and_pepper}"
+        return repr
+
+    def make_samplers(self, device: torch.device, dtype: torch.dtype) -> None:
+        r"""Create samplers for generating random noise parameters."""
+        amount = _range_bound(
+            self.amount,
+            "amount",
+        ).to(device)
+        salt_and_pepper = _range_bound(
+            self.salt_and_pepper,
+            "salt_and_pepper",
+        ).to(device)
+
+        self.amount_sampler = UniformDistribution(amount[0], amount[1], validate_args=False)
+        self.salt_and_pepper_sampler = UniformDistribution(salt_and_pepper[0], salt_and_pepper[1], validate_args=False)
+
+    def forward(self, batch_shape: tuple[int, ...], same_on_batch: bool = False) -> dict[str, Tensor]:
+        r"""Generate Salt and Pepper noise masks for a batch of images."""
+        batch_size, C, H, W = batch_shape
+        _common_param_check(batch_size, same_on_batch)
+        _device, _dtype = _extract_device_dtype([self.amount, self.salt_and_pepper])
+
+        amount_factor = _adapted_rsampling((batch_size,), self.amount_sampler, same_on_batch).to(
+            device=_device, dtype=_dtype
+        )
+        salt_and_pepper_factor = _adapted_rsampling((batch_size,), self.salt_and_pepper_sampler, same_on_batch).to(
+            device=_device, dtype=_dtype
+        )
+
+        ## Generate noise masks.
+        mask_noise = _adapted_uniform(
+            (batch_size, 1, H, W), low=0.0, high=1.0, same_on_batch=same_on_batch
+        ) < amount_factor.view(batch_size, 1, 1, 1)
+        mask_salt = _adapted_uniform(
+            (batch_size, 1, H, W), low=0.0, high=1.0, same_on_batch=same_on_batch
+        ) < salt_and_pepper_factor.view(batch_size, 1, 1, 1)
+
+        # If the number of channels is greater than one (3), replicate the generated mask for each channel.
+        if C > 1:
+            mask_noise = mask_noise.repeat(1, C, 1, 1)
+            mask_salt = mask_salt.repeat(1, C, 1, 1)
+        mask_pepper = (~(~mask_salt & mask_noise)).float()
```

Comment:
Yes, we should. It is designed as this since the different random state between CPU and GPU. If the random generator is on GPU, then all the numerical tests need to be updated to have two versions.

Reply:
I have modified it as follows:
mask_pepper = (~(~mask_salt & mask_noise)).to(device=_device, dtype=_dtype)
(_device, _dtype) --> Extracted from --> _extract_device_dtype([self.amount, self.salt_and_pepper])